src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1169,13 +1143,13 @@ private FileStreamAsyncResult BeginReadAsync(byte[] array, int offset, int numBy
                     Buffer.BlockCopy(_buffer, _readPos, array, offset, n);
                     _readPos += n;
 
-                    // Return a synchronous FileStreamAsyncResult
-                    return FileStreamAsyncResult.CreateBufferedReadResult(n, userCallback, stateObject, false);
+                    // Return a completed task
+                    return Task.FromResult(n);
```

Comment:
For cases where we can successfully fill the request based on the buffer, we should consider trying to use a cached task.  You can maintain a ```Task<int> _lastSuccessfulRead``` on the instance.  Then here, you can do something like:
```
Task<int> lastSuccessfulRead = _lastSuccessfulRead;
return lastSuccessfulRead != null && lastSuccessfulRead.Result == n ?
    lastSuccessfulRead :
    (_lastSuccessfulRead = Task.FromResult(n));
```
Since it's very likely that the user is calling ReadAsync with the same length each time, it's quite likely that the last successful task will have had the same length, in which case we can reuse it.